### PR TITLE
Add support for Inferentia ASICs

### LINKED ISSUE #654: Support inferentia instances
#### Description

* https://aws.amazon.com/ec2/instance-types/inf1/
* inferentia [blog post](https://aws.amazon.com/blogs/aws/amazon-ec2-update-inf1-instances-with-aws-inferentia-chips-for-high-performance-cost-effective-inferencing/), [announcement](https://aws.amazon.com/about-aws/whats-new/2019/12/introducing-amazon-ec2-inf1-instances-high-performance-and-the-lowest-cost-machine-learning-inference-in-the-cloud/)

### PR DESCRIPTION
**Pull request by RobertLucian**
Closes #654.
It's a continuation of #1071.

Before merge TODOs:

- [x] pattern-match `python-predictor-inf/Dockerfile` with `python-predictor-gpu/Dockerfile`
- [x] update `start_uvicorn.py` for multi-model inferentia
- [x] add validation to return error if trying to use inf on local
- [x] add validation to disallow local paths on aws - was already supported
- [x] reference inferentia examples in `inferentia.md`
---

Merge checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [x] update examples
- [x] update docs and add any new files to `summary.md` (view in [gitbook](https://docs.cortex.dev/v/master) after merging)

### CODE REVIEW

File: pkg/workloads/cortex/lib/storage/concurrency.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    def acquire(self):
+        """
+        To acquire rw access to resource.
+        """
+        if self._file_handle:
+            return
+
+        if not self.timeout:
+            self._file_handle = open(self._lock_file, "w")
```
[the remark this answers is not in the document]
Comment (deliahu) [reply pkg/workloads/cortex/lib/storage/concurrency.py:44]: After discussion, we determined it is not safe to use one instance of this object in multiple threads. acquire() and release() must be called sequentially (unless we use threading locks). Do you mind adding this info to the docstring?

File: dev/versions.md [hunk 2]
```diff
@@ -163,11 +167,28 @@ Note: it's ok if example training notebooks aren't upgraded, as long as the expo
...
 
 Note: overriding horizontal-pod-autoscaler-sync-period on EKS is currently not supported (<https://github.com/awslabs/amazon-eks-ami/issues/176>)
 
+## Neuron RTD
+
+1. Run a `cortexlabs/neuron-rtd` container and check if there are newer versions of `aws-neuron-tools` and `aws-neuron-runtime` with `yum info <package>` command.
+1. Add in the newer versions inside of `images/neuron-rtd/Dockerfile`.
+1. Match the neuron versions in `images/python-predictor-inf/Dockerfile` with those of the Neuron RTD.
+1. Match the neuron versions in `images/tensorflow-serving-inf/Dockerfile` with those of the Neuron RTD.
+1. Rebuild all afferent images.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply dev/versions.md:181]: In this case, it refers to the 2 above images: `images/python-predictor-inf/Dockerfile` and `images/tensorflow-serving-inf/Dockerfile`.

File: docs/deployments/compute.md [hunk 3]
```diff
@@ -15,18 +15,22 @@ For example:
...
 
 See [GPU documentation](gpus.md) for more information.
+
+## Memory
+
+One unit of memory is one byte. Memory can be expressed as an integer or by using one of these suffixes: `K`, `M`, `G`, `T` (or their power-of two counterparts: `Ki`, `Mi`, `Gi`, `Ti`). For example, the following values represent roughly the same memory: `128974848`, `129e6`, `129M`, `123Mi`.
+
+## Inf
+
+One unit of Inf corresponds to one Inferentia ASIC with 4 NeuronCores and 8GB of cache memory. Fractional requests are not allowed.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply docs/deployments/compute.md:36]: Mentioning the number of available *NeuronCores* and available *cache memory* is required for the understanding of what an Inferentia ASIC offers. Otherwise, the users can waste time trying to get to understand that - the AWS documentation is not excellent either, so this way, we speed up the process of getting it. We can probably attach a hyperlink to *"Inferentia"* to make it obvious that we refer to that. Or we can say that *NeuronCores* and the *cache memory* is not the same thing as `cpu` and `mem`. What do you say?

Also, the cache memory of the Inferentia ASIC is in a way an extension of the system RAM memory.

File: docs/deployments/inferentia.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# Using Inferentia
+
+_WARNING: you are on the master branch, please refer to the docs on the branch that matches your `cortex version`_
+
+To use [Inferentia ASICs](https://aws.amazon.com/machine-learning/inferentia/):
+
+1. You may need to [file an AWS support ticket](https://console.aws.amazon.com/support/cases#/create?issueType=service-limit-increase&limitType=ec2-instances) to increase the limit for your desired instance type.
+1. Set the instance type to an AWS Inferentia instance (e.g. `inf1.xlarge`) when creating your Cortex cluster.
+1. Set the `inf` field in the `compute` configuration for your API. One unit of `inf` corresponds to one Inferentia ASIC with 4 NeuronCores and 8GB of cache memory. Fractional requests are not allowed.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply docs/deployments/inferentia.md:9]: Same as above.

File: docs/deployments/inferentia.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+## Neuron
+
+Inferentia ASICs come in different sizes depending on the instance type:
+
+* `inf1.xlarge`/`inf1.2xlarge` - each has 1 Inferentia ASIC
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply docs/deployments/inferentia.md:15]: I think it would be better to keep it as `Inferentia ASIC`. I saw that the AWS documentation refers to them as both Inferentia chips and ASICs. Calling them chips is not specific enough, it's too generic. And other providers such as Google already call them [ASICs (TPU)](https://en.wikipedia.org/wiki/Tensor_processing_unit). So, calling them by what they are (ASICs and not chips) better suites the scheme as we keep adding support for other ASICs.

Chips are CPUs too, RAM modules too, GPUs as well, SSDs as well - basically, any embedded electronic device is a chip. It's just not specific enough. ASICs on the other hand, refer to something *S* (specifically) built for some kind of application, which in this case, it's a chip specialized for ML inference.

File: examples/pytorch/image-classifier-resnet50/cortex_cpu.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      model_name: resnet50.pt
+      device: cpu
+      classes: https://s3.amazonaws.com/deep-learning-models/image-models/imagenet_class_index.json
+      input_shape: [224, 224]
+  compute:
+    cpu: 2
+    mem: 2G
+  autoscaling:
+    workers_per_replica: 2
+    threads_per_worker: 4
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/pytorch/image-classifier-resnet50/cortex_cpu.yaml:18]: No particular reason, other than it can handle that many threads without being bottlenecked. Do you think we should take this out?

File: examples/pytorch/image-classifier-resnet50/cortex_gpu.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+      device: gpu
+      classes: https://s3.amazonaws.com/deep-learning-models/image-models/imagenet_class_index.json
+      input_shape: [224, 224]
+  compute:
+    gpu: 1
+    cpu: 2
+    mem: 2G
+  autoscaling:
+    workers_per_replica: 2
+    threads_per_worker: 1
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/pytorch/image-classifier-resnet50/cortex_gpu.yaml:19]: I'm thinking that even though that value is set by default, it is easier for the user to understand that there's one thread per worker. This is to spare him the effort of having to go through the documentation. What do you think? Do you think we should take this out?

File: examples/pytorch/image-classifier-resnet50/cortex_inf.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+      device: inf
+      classes: https://s3.amazonaws.com/deep-learning-models/image-models/imagenet_class_index.json
+      input_shape: [224, 224]
+  compute:
+    inf: 1
+    cpu: 2
+    mem: 2G
+  autoscaling:
+    workers_per_replica: 2
+    threads_per_worker: 1
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/pytorch/image-classifier-resnet50/cortex_inf.yaml:19]: Same as above.

File: examples/pytorch/image-classifier-resnet50/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+```bash
+curl <API endpoint> -X POST -H "Content-Type: application/json" -d @sample.json
+```
+
+This image is embedded in [sample.json](sample.json):
+
+![image](https://i.imgur.com/213xcvs.jpg)
+
+If a 5-element list is returned containing classifications of the image ("tabby", "Egyptian_cat", "tiger_cat", "tiger", "plastic_bag"), then it means the API is working. The first classification in the list is the most likely.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/pytorch/image-classifier-resnet50/README.md:37]: Got it, will change this.

File: examples/tensorflow/image-classifier-resnet50/README.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+Before [throughput_test.py](throughput_test.py) is run, 2 environment variables have to be exported:
+
+```bash
+export ENDPOINT=<API endpoint>  # you can find this with `cortex get image-classifier-resnet50`
+export IMG_URL=https://i.imgur.com/213xcvs.jpg # this is the cat image shown in the previous step
+```
+
+Then, deploy each API one at a time and check the results:
+
+1. Running `python throughput_test.py -i 30 -w 4 -t 48` with the [cortex_inf.yaml](cortex_inf.yaml) API running on an `inf1.2xlarge` instance will get **~510 inferences/sec** with an average latency of **80 ms**.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/tensorflow/image-classifier-resnet50/README.md:74]: We want to show how much performance you can get per dollar with Inferentia ASICs. The only instance that makes sense economically is `inf1.2xlarge`, which is almost twice as cost-effective as `inf1.xlarge`. That's because the major bottleneck is with the CPU and not with the Inferentia ASIC. If the CPU were to be more performant, the throughput could theoretically be as almost as 8 times as higher (at least in this example).

Basically, when the throughput is high, the costs with `inf1.xlarge` are high too, but with `inf1.2xlarge` they are low.

File: examples/tensorflow/image-classifier-resnet50/README.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+```bash
+curl <API endpoint> -X POST -H "Content-Type: application/json" -d @sample.json
+```
+
+This image is embedded in [sample.json](sample.json):
+
+![image](https://i.imgur.com/213xcvs.jpg)
+
+If a 5-element list is returned containing classifications of the image ("tabby", "Egyptian_cat", "tiger_cat", "tiger", "plastic_bag"), then it means the API is working. The first classification in the list is the most likely.
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/tensorflow/image-classifier-resnet50/README.md:37]: Yep, copy that. Will do it.

File: pkg/types/spec/validations.go [hunk 12]
```diff
@@ -823,6 +852,20 @@ func isValidTensorFlowS3Directory(path string, awsClientForBucket *aws.Client) b
 	return true
 }
 
+// isValidNeuronTensorflowS3Directory checks that the path contains a valid S3 directory for Neuron TensorFlow models
+// Must contain the following structure:
+// - 1523423423/ (version prefix, usually a timestamp)
+// 		- saved_model.pb
+func isValidNeuronTensorflowS3Directory(path string, awsClient *aws.Client) bool {
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply pkg/types/spec/validations.go:859]: Good catch! Thanks!

File: pkg/lib/math/int.go [hunk 13]
```diff
@@ -35,3 +35,17 @@ func MaxInt(val int, vals ...int) int {
...
 	return max
 }
+
+func IsDivisibleByInt(num int, divisor int) bool {
+	return num%divisor == 0
+}
+
+func FactorsInt(num int) []int {
+	divisibleNumbers := []int{}
+	for divisor := 1; divisor <= num/2; divisor++ {
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply pkg/lib/math/int.go:45]: Funny thing, this is what David and I thought too the first time. Getting the square root is only necessary for determining if a number is prime or not. Dividing it in half is what's needed when finding out the factors of a number (2 can be a factor of it too).

[the remark this answers is not in the document]
Comment (deliahu) [reply pkg/lib/math/int.go:45]: Haha yes we got confused by this too. Now that I think about it more, checking only up to the square root would work if we add the other factor to the list as we go, but if we only check the mod without dividing (as it's written now) it's necessary to check up to `num/2` (as you said)

File: docs/deployments/compute.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (RobertLucian) [reply docs/deployments/compute.md:36]: Specify that this isn't the same thing as `cpu` and `mem`.

File: examples/pytorch/image-classifier-resnet50/cortex_cpu.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/pytorch/image-classifier-resnet50/cortex_cpu.yaml:18]: Remove `autoscaling` section.

File: examples/tensorflow/image-classifier-resnet50/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (RobertLucian) [reply examples/tensorflow/image-classifier-resnet50/README.md:74]: Add reason why `inf1.xlarge` wasn't used in this example.

**Comment by RobertLucian** [comment]
Ran the following tests to double-check everything is still okay:

1. Tested every Cortex deployment that uses Inferentia ASICs. 
1. Tested the multi-model examples.
1. Tested some generic single-model examples (PyTorch, TensorFlow, ONNX).
1. Verified that the CloudWatch dashboard works.
1. Converted an Inferentia-provisioned example to use multiple models with the TF Predictor and tested it. This example is not present in the PR per se.
1. Checked that the Autoscaler spins up/down new/old nodes when the situation asks for it.
